### PR TITLE
`RPA.Email.Exchange` OAuth2 support

### LINKED ISSUE #567: `RPA.Email.Exchange` Implement Oauth2 flow as basic auth is being deprecated
Per the top of the [Exchange Web Services API Overview page](https://docs.microsoft.com/en-us/exchange/client-developer/exchange-web-services/get-started-with-ews-managed-api-client-applications), Microsoft is deprecating basic authentication by October 2022. We need to implement the Oauth2 flow in its place. The underlying `credentials.py` appears to already have a class for Oauth2.

### PR DESCRIPTION
**Pull request by cmin764**
Closes #567 
Working Portal [robot](https://github.com/robocorp/example-oauth-email/blob/feature/microsoft-oauth2/tasks.robot#L73-L79)

### ToDo
- [x] Optionally add a robot test (for displaying how the flow works)
  - note that the unit-test is already skipped even for the basic auth case

### CODE REVIEW

File: packages/main/src/RPA/Email/Exchange.py [hunk 1]
```diff
@@ -223,13 +255,37 @@ def authorize(
         kwargs["primary_smtp_address"] = (
             primary_smtp_address if primary_smtp_address else username
         )
-        self.credentials = Credentials(username, password)
+        if is_oauth:
+            self.credentials = OAuth2Creds(
+                on_token_refresh=self.on_token_refresh,
```
Comment (blakewaud) [inline packages/main/src/RPA/Email/Exchange.py:260]: I'm guessing there's some code I'm not following in the `oauthlib` library, but from what I see, we pass in a function that only logs to override the auto refresh method. Assuming I am wrong about this, you can ignore my comment and I'd appreciate learning how this is actually working.

**Review by blakewaud — APPROVED** [review]
LGTM! One question on the refresh functionality likely for my own understanding.

File: packages/main/src/RPA/Email/Exchange.py (hunk 1 above)
Comment (cmin764) [reply packages/main/src/RPA/Email/Exchange.py:260]: You're right 100% and at first glance this seems useless, but it actually adds support for customizing token refresh behaviour with ease in your robot, like in this [example](https://github.com/robocorp/example-oauth-email/pull/1/files#diff-b93aca6dec83011aab894b95b6deaaf19b48d5870212c8366f1e0e66acd3cc80R29-R33):
```python
class ExtendedExchange(Exchange):

    def __init__(self, *args, secret_name, **kwargs):
        super().__init__(*args, **kwargs)
        self._secret_name = secret_name

    def on_token_refresh(self, token):
        super().on_token_refresh(token)
        secret = vault.get_secret(self._secret_name)
        secret["token"] = dict(token)
        vault.set_secret(secret)
```
which saves the token in the vault on refresh so the new one can be picked up next time the robot decides to run again to authorize and send an e-mail.

Comment (blakewaud) [reply packages/main/src/RPA/Email/Exchange.py:260]: That's great... and now I'm thinking your comment should be added to the documentation as an advanced use case somehow! 😄

Comment (cmin764) [reply packages/main/src/RPA/Email/Exchange.py:260]: Yes, not yet in the docs but exists in the docstring of the `on_token_refresh` method and most of the instructions will go in the Portal [bot](https://robocorp.com/portal/robot/robocorp/example-oauth-email) which show-cases the extension of `Exchange`. I need at least one customer dealing with OAuth2 flow on `RPA.Email.Exchange` through the library before committing any more focus into this. (as I bet most of them will still prefer or be able to use just "App Passwords")

**Pull request merged**